### PR TITLE
draft for sdk-docs

### PR DESCRIPTION
**Pull request by imLew**
**Proposed changes**:
- Add a docs page that describes the components of rasa-sdk.
So far it only contains a description of attributes of `Tracker` and a copy of the custom action docs.

To able to use `automethod` some docstrings were also updated in the sdk https://github.com/RasaHQ/rasa-sdk/pull/128

**Status (please check what you already did)**:
- [ ] made PR ready for code review
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

**Comment by erohmensing** [comment]
I would add information about the events -- how to return them, what types there are, etc.

**Comment by tmbo** [comment]
@imLew can we get this to a version that we can merge? want to avoid that this is getting stale

File: docs/api/rasa-sdk.rst [hunk 1]
```diff
@@ -0,0 +1,90 @@
+:desc: Information about `rasa-sdk` objects and functions.
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]: Can you break out your inner marketer to try to fancy this up? It's for SEO optimization so something like "Customize your Rasa assistant and connect to external APIs using the Rasa-SDK package to write custom actions in Python" or something like that 😄

File: docs/api/rasa-sdk.rst [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+.. _rasa-sdk:
+
+Rasa SDK
+========
+
+.. edit-link::
+
+Rasa SDK provides the tools you need to write :ref:`custom-actions`.
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]:
```suggestion
Rasa SDK provides the tools you need to write :ref:`custom-actions` in Python.
```

Comment (erohmensing) [reply docs/api/rasa-sdk.rst]: Can we add a quick sentence about the sdk having a few relevant classes to action-writing to lead into contents, since the contents list is currently a list of classes?

File: docs/api/rasa-sdk.rst [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+.. contents::
+   :local:
+
+``Action``
+----------
+
+The ``Action`` class is the base class for any custom action. It has two methods
+that both need to be overwritten, `.name()` and `.run()`.
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]:
```suggestion
that both need to be overwritten, ``name()`` and ``run()``.
```

File: docs/api/rasa-sdk.rst [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+.. _custom_action_example_verbose:
+
+In a restaurant bot, if the user says "show me a Mexican restaurant",
+your bot could execute the action ``ActionCheckRestaurants``,
+which might look like this:
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]: maybe mention that it is connecting to an external database

File: docs/api/rasa-sdk.rst [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+         return "action_check_restaurants"
+
+      def run(self,
+              dispatcher: CollectingDispatcher,
+              tracker: Tracker,
+              domain: Dict[Text, Any]) -> List[Dict[Text, Any]]:
+
+         cuisine = tracker.get_slot('cuisine')
+         q = "select * from restaurants where cuisine='{0}' limit 1".format(cuisine)
+         result = db.query(q)
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]: Is this just pseudocode or was it taken somehwere? is there an import for `db`?

File: docs/api/rasa-sdk.rst [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+         return [SlotSet("matches", result if result is not None else [])]
+
+
+You should add the the action name ``action_check_restaurants`` to
+the actions in your domain file. The action's ``run`` method receives
+three arguments. You can access the values of slots and the latest message
+sent by the user using the ``tracker`` object, and you can send messages
+back to the user with the ``dispatcher`` object, by calling
+``dispatcher.utter_template``, ``dispatcher.utter_message``, or any other
+``rasa_sdk.executor.CollectingDispatcher`` method.
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]: "by calling `dispatcher.utter_message`" now!

Maybe we can actually put more information about `utter_message` and all of the different keyword arguments it takes? could even do it via automethod probably

File: docs/api/rasa-sdk.rst [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+.. automethod:: rasa_sdk.Action.run
+   :noindex:
+
+``Events``
+----------
+
+An action's ``run`` method returns a list of events. For more information on
+events see :ref:`Events`. The action itself will automatically be added to the
+tracker as an ``ActionExecuted`` event. If the action should not trigger any
+other events it must return an empty list.
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]:
```suggestion
other events, it must return an empty list.
```

Comment (erohmensing) [reply docs/api/rasa-sdk.rst]: not strictly true, but good that we encourage good practice 😄

File: docs/api/rasa-sdk.rst [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+``Tracker``
+-----------
+
+The ``rasa_sdk.Tracker`` lets you access the bots memory in your custom
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]:
```suggestion
The ``rasa_sdk.Tracker`` lets you access the bot's memory in your custom
```

File: docs/api/rasa-sdk.rst [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+In no particular order the attributes of the ``Tracker`` are:
+
+- ``sender_id`` - The unique ID of person talking to the bot.
+- ``slots`` - The list of slots that can be filled as defined in the
+  "ref"`domains`.
+- ``latest_message`` - A dictionary containing the attributes of the latest
+  message: ``intent``, ``entities`` and ``text``.
+- ``latest_event_time`` - The timestap from the last event that occured.
+- ``paused`` - Whether the tracker is currently paused.
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]: from the code it looks like that rather lives at `_paused`

File: docs/api/rasa-sdk.rst [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+- ``events`` - A list of all previous events.
+- ``latest_input_channel`` - The name of the input channel of the last
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]: don't think that is available as an attribute, just via `get_latest_input_channel()`

File: docs/api/rasa-sdk.rst [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+conversation through ``Tracker`` attributes.
+
+In no particular order the attributes of the ``Tracker`` are:
+
+- ``sender_id`` - The unique ID of person talking to the bot.
+- ``slots`` - The list of slots that can be filled as defined in the
+  "ref"`domains`.
+- ``latest_message`` - A dictionary containing the attributes of the latest
+  message: ``intent``, ``entities`` and ``text``.
+- ``latest_event_time`` - The timestap from the last event that occured.
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]: Also not available as a property

File: docs/api/rasa-sdk.rst [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+- ``paused`` - Whether the tracker is currently paused.
+- ``events`` - A list of all previous events.
+- ``latest_input_channel`` - The name of the input channel of the last
+  ``UserUttered`` event.
+- ``active_form`` - The name of the currently active form.
+- ``latest_action_name`` - The name of the last action the bot executed.
+
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]: Let's talk about the methods as well that may be relevant, like `get_slot` and `get_latest_entity_values`

File: docs/api/rasa-sdk.rst (hunk 1 above)
Comment (imLew) [reply docs/api/rasa-sdk.rst]: Didn't know this was for SEO
😂 I'll try to make it flashy, add all the buzzwords I can think of

File: docs/api/rasa-sdk.rst (hunk 5 above)
Comment (imLew) [reply docs/api/rasa-sdk.rst]: This was copied from the `Actions` page

Comment (erohmensing) [reply docs/api/rasa-sdk.rst]: eh, alright haha

File: docs/api/rasa-sdk.rst (hunk 6 above)
Comment (imLew) [reply docs/api/rasa-sdk.rst]: Updated that :P 
But automethod does not seem to get the latest rasa-sdk, at least locally even if I pip install the sdk from a local copy 😅

File: docs/api/rasa-sdk.rst (hunk 10 above)
Comment (imLew) [reply docs/api/rasa-sdk.rst]: Fixed all these, it only lists actual attributes now and uses automethod on the tracker methods. I will be adding a small PR to SDK to improve the docstrings for this.

**Comment by imLew** [comment]
@erohmensing incorporated your review and moved the 'custom actions in python' section as we discussed, please have another final look before we merge

**Comment by erohmensing** [comment]
uh, this looks like a failed rebase or something? was master merged into this branch?

**Comment by tmbo** [comment]
given that there is a minor release scheduled next week anyways, let's merge this into master, it is easier now in any case :D

**Comment by tmbo** [comment]
changed base, @erohmensing can you give this another review, please?

File: docs/api/rasa-sdk.rst [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+You should add the the action name ``action_check_restaurants`` to
+the actions in your domain file. The action's ``run`` method receives
+three arguments. You can access the values of slots and the latest message
+sent by the user using the ``tracker`` object, and you can send messages
+back to the user with the ``dispatcher`` object, by calling
+``dispatcher.utter_message``.
+
+Details of the ``run()`` method:
+
+.. automethod:: rasa_sdk.Action.run
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]: the `returns:` part of this renders weirdly, can you take a look at the docstring for that?

File: docs/api/rasa-sdk.rst [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+Running the Action Server
+-------------------------
+
+If you have ``rasa`` installed, run this command to start your action server:
+
+.. code-block:: bash
+
+    rasa run actions
+
+.. _custom_action_example:
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]: why is this here?

File: docs/api/rasa-sdk.rst [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+Details of the ``dispatcher.utter_message()`` method:
+
+.. automethod:: rasa_sdk.executor.CollectingDispatcher.utter_message
+
+Events
+------
+
+An action's ``run()`` method returns a list of events. For more information on
+the different types of events, see :ref:`Events`. There is an example of a ``SlotSet`` event
+:ref:`above <custom_action_example>`. The action itself will automatically be added to the
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]: i think that is supposed to reference `custom_action_example_verbose`

File: docs/api/rasa-sdk.rst [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+back to the user with the ``dispatcher`` object, by calling
+``dispatcher.utter_message``.
+
+Details of the ``run()`` method:
+
+.. automethod:: rasa_sdk.Action.run
+
+Details of the ``dispatcher.utter_message()`` method:
+
+.. automethod:: rasa_sdk.executor.CollectingDispatcher.utter_message
```
Comment (erohmensing) [inline docs/api/rasa-sdk.rst]: docstring on this one renders a bit weird as well

File: docs/api/rasa-sdk.rst (hunk 14 above)
Comment (imLew) [reply docs/api/rasa-sdk.rst]: the docs would be very different had you been here from the beginning to review them xP

File: docs/api/rasa-sdk.rst (hunk 13 above)
Comment (imLew) [reply docs/api/rasa-sdk.rst]: done

**Comment by imLew** [comment]
@erohmensing fixes for the docstrings are here https://github.com/RasaHQ/rasa-sdk/pull/128, feel free to review/add ideas :D

File: docs/api/rasa-sdk.rst (hunk 14 above)
Comment (erohmensing) [reply docs/api/rasa-sdk.rst]: Hahahaha

**Comment by erohmensing** [comment]
Coolio, I'll take a look, don't let that stop you from merging this one 👍
